src/HTTP/Psr18Client.php

```diff
@@ -0,0 +1,144 @@
+ */
+final class Psr18Client implements Client
+{
+    private $httpClient;
+
+    private $requestFactory;
+
+    private $uriFactory;
+
+    private $allowedRedirects = 5;
+
+    public function __construct(ClientInterface $httpClient, RequestFactoryInterface $requestFactory, UriFactoryInterface $uriFactory)
+    {
+        $this->httpClient = $httpClient;
+        $this->requestFactory = $requestFactory;
+        $this->uriFactory = $uriFactory;
+    }
+
+    public function getHttpClient(): ClientInterface
+    {
+        return $this->httpClient;
+    }
+
+    public function getRequestFactory(): RequestFactoryInterface
+    {
+        return $this->requestFactory;
+    }
+
+    public function getUriFactory(): UriFactoryInterface
+    {
+        return $this->uriFactory;
+    }
+
+    /**
+     * send a request and return the response
+     *
+     * @param Client::METHOD_GET $method
+     * @param string $url
+     * @param string[] $headers
+     *
+     * @throws HttpException if anything goes wrong requesting the data
+     */
+    public function request(string $method, string $url, array $headers = []): Response
+    {
+        if ($method !== self::METHOD_GET) {
+            throw new InvalidArgumentException(sprintf(
+                '%s(): Argument #1 ($method) only supports method "%s".',
+                __METHOD__,
+                self::METHOD_GET
+            ), 1);
+        }
+
+        if (preg_match('/^http(s)?:\/\//i', $url)) {
+            return $this->requestUrl($method, $url, $headers);
+        }
+
+        return $this->requestLocalFile($url);
+    }
+
+    private function requestUrl(string $method, string $url, array $headers): Response
+    {
+        $permanentUrl = $url;
+        $requestedUrl = $url;
+        $remainingRedirects = $this->allowedRedirects;
+
+        $request = $this->requestFactory->createRequest(
+            $method,
+            $this->uriFactory->createUri($requestedUrl)
+        );
+
+        foreach ($headers as $name => $value) {
+            $request = $request->withHeader($name, $value);
+        }
+
+        do {
+            $followRedirect = false;
+
+            try {
+                $response = $this->httpClient->sendRequest($request);
+            } catch (ClientExceptionInterface $th) {
+                throw new HttpException($th->getMessage(), $th->getCode(), $th);
+            }
+
+            $statusCode = $response->getStatusCode();
+
+            // If we have a redirect
+            if (in_array($statusCode, [300, 301, 302, 303, 307]) && $response->hasHeader('Location')) {
+                // Prevent infinity redirect loops
+                if ($remainingRedirects <= 0) {
+                    break;
+                }
+
+                $remainingRedirects--;
+                $followRedirect = true;
+
+                $requestedUrl = $response->getHeaderLine('Location');
+
+                if ($statusCode === 301) {
+                    $permanentUrl = $requestedUrl;
```

Comment:
We need some kind of mechanism to not change the permanent URL after temporary redirects (as tested by https://github.com/simplepie/simplepie/pull/819). For example, this is what File does:

https://github.com/simplepie/simplepie/blob/e356f5dc711ad4c5d92091a7dea509bf7c1c11a6/src/File.php#L144-L144